extmap.go

```diff
@@ -0,0 +1,88 @@
+package sdp
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/pions/webrtc/pkg/rtcerr"
+
+	"github.com/pkg/errors"
+)
+
+//ExtMap represents the activation of a single RTP header extension
+type ExtMap struct {
+	Value     int
+	Direction Direction
+	URI       *url.URL
+	ExtAttr   *string
+}
+
+//Clone converts this object to an Attribute
+func (e *ExtMap) Clone() Attribute {
+	return Attribute{Key: "extmap", Value: e.string()}
+}
+
+//Unmarshal creates an Extmap from a string
+func (e *ExtMap) Unmarshal(raw string) error {
+	parts := strings.SplitN(raw, ":", 2)
+	if len(parts) != 2 {
+		return errors.Wrap(&rtcerr.SyntaxError{Err: fmt.Errorf("%v", raw)}, pkgName)
+	}
+
+	fields := strings.Fields(parts[1])
+	if len(parts) < 2 {
+		return errors.Wrap(&rtcerr.SyntaxError{Err: fmt.Errorf("%v", raw)}, pkgName)
+	}
+
+	valdir := strings.Split(fields[0], "/")
+	value, err := strconv.ParseInt(valdir[0], 10, 64)
+	if err != nil {
+		return errors.Wrap(&rtcerr.SyntaxError{Err: fmt.Errorf("%v", valdir[0])}, pkgName)
+	}
+
+	var direction Direction
+	if len(valdir) == 2 {
+		direction = NewDirection(valdir[1])
+	}
+
+	uri, err := url.Parse(fields[1])
+	if err != nil {
+		return errors.Wrap(&rtcerr.UnknownError{Err: err}, pkgName)
+	}
+
+	if len(fields) == 3 {
+		tmp := fields[2]
+		e.ExtAttr = &tmp
+	}
+
+	e.Value = int(value)
+	e.Direction = direction
+	e.URI = uri
+	return nil
+}
+
+//Marshal creates a string from an ExtMap
+func (e *ExtMap) Marshal() string {
+	return attributeKey + e.Name() + ":" + e.string() + endline
+}
+
+func (e *ExtMap) string() string {
+	output := fmt.Sprintf("%d", e.Value)
+	if e.Direction != Direction(unknown) {
+		output += "/" + e.Direction.String()
+	}
+
+	output += " " + e.URI.String()
```

Comment:
When E.URI is nil this will panic

Reply:
Not if the library is used by design because there is unmarshaling during the Unmarshal function. The only situation this may occur under is when someone creates `ExtMap{}` externally. In which case I am not entirely sure what someone would do that.

I guess maybe we could add checks? Not sure what should occur if the check fails. Since this is a `Stringer` interface function which must return a string, if this fails should it return an empty string? I would have said that given that this is a parser this kind of failure should be detected in the parsing process to guarantee that `String()` works.

I am writing in circles. Anyway I think you guys get my drift here.

Reply:
hmmm...in this case should string() return an empty string, or a malformed one?

Reply:
@wdouglass I have no idea. Your call here. But I refer to the point I mentioned in above comment. That's my personal design approach I take generally.